src/library/scala/concurrent/impl/FutureConvertersImpl.scala

```diff
@@ -12,19 +12,84 @@
 
 package scala.concurrent.impl
 
-import java.util.concurrent.{CompletableFuture, CompletionStage}
-import java.util.function.BiConsumer
+import java.util.concurrent.{CompletableFuture, CompletionStage, TimeUnit}
+import java.util.function.{BiConsumer, BiFunction, Consumer, Function => JFunction}
 
 import scala.concurrent.Future
 import scala.concurrent.impl.Promise.DefaultPromise
 import scala.util.{Failure, Success, Try}
 
-object FutureConvertersImpl {
+private[scala] object FutureConvertersImpl {
   final class CF[T](val wrapped: Future[T]) extends CompletableFuture[T] with (Try[T] => Unit) {
     override def apply(t: Try[T]): Unit = t match {
       case Success(v) => complete(v)
       case Failure(e) => completeExceptionally(e)
     }
+
+    // Ensure that completions of this future cannot hold the Scala Future's completer hostage
+
+    override def thenApply[U](fn: JFunction[_ >: T, _ <: U]): CompletableFuture[U] = thenApplyAsync(fn)
+
+    override def thenAccept(fn: Consumer[_ >: T]): CompletableFuture[Void] = thenAcceptAsync(fn)
+
+    override def thenRun(fn: Runnable): CompletableFuture[Void] = thenRunAsync(fn)
+
+    override def thenCombine[U, V](cs: CompletionStage[_ <: U], fn: BiFunction[_ >: T, _ >: U, _ <: V]): CompletableFuture[V] = thenCombineAsync(cs, fn)
+
+    override def thenAcceptBoth[U](cs: CompletionStage[_ <: U], fn: BiConsumer[_ >: T, _ >: U]): CompletableFuture[Void] = thenAcceptBothAsync(cs, fn)
+
+    override def runAfterBoth(cs: CompletionStage[_], fn: Runnable): CompletableFuture[Void] = runAfterBothAsync(cs, fn)
+
+    override def applyToEither[U](cs: CompletionStage[_ <: T], fn: JFunction[_ >: T, U]): CompletableFuture[U] = applyToEitherAsync(cs, fn)
+
+    override def acceptEither(cs: CompletionStage[_ <: T], fn: Consumer[_ >: T]): CompletableFuture[Void] = acceptEitherAsync(cs, fn)
+
+    override def runAfterEither(cs: CompletionStage[_], fn: Runnable): CompletableFuture[Void] = runAfterEitherAsync(cs, fn)
+
+    override def thenCompose[U](fn: JFunction[_ >: T, _ <: CompletionStage[U]]): CompletableFuture[U] = thenComposeAsync(fn)
+
+    override def whenComplete(fn: BiConsumer[_ >: T, _ >: Throwable]): CompletableFuture[T] = whenCompleteAsync(fn)
```

Comment:
@lrytz @SethTisue @viktorklang Why all these methods are delegated to the async ones?

Reply:
There is a comment above:

```
    // Ensure that completions of this future cannot hold the Scala Future's completer hostage
```

It seems these overrides were always there https://github.com/scala/scala-java8-compat/commit/b3fca14e9ffe920b0b1ebb9f9f0f104c0869b6fc

Haven't thought about it in detail, but maybe there are separate thread pools for Scala and Java futures, and it's to avoid running on the wrong one?

Reply:
Amazingly, I could scroll back three weeks in Discord and find the link I posted.

https://github.com/scala/scala-java8-compat/blob/main/src/main/scala/scala/compat/java8/FutureConverters.scala#L57